wallet-new/src/Cardano/Wallet/Types/UtxoStatistics.hs

```diff
@@ -0,0 +1,191 @@
+import           Data.Aeson
+import qualified Data.HashMap.Strict as HMS
+import qualified Data.List.NonEmpty as NL
+import qualified Data.Map.Strict as Map
+import           Data.Scientific (floatingOrInteger)
+import           Data.Swagger hiding (Example, example)
+import qualified Data.Text as T
+import           Data.Word (Word64)
+import           Formatting (bprint, build, (%))
+import qualified Formatting.Buildable
+import           Serokell.Util (listJson)
+import           Test.QuickCheck
+
+import           Cardano.Wallet.API.V1.Swagger.Example (Example)
+import           Pos.Infra.Util.LogSafe (BuildableSafeGen (..),
+                     deriveSafeBuildable)
+
+
+--  Utxo statistics for the wallet.
+--  Histogram is composed of bars that represent the bucket. The bucket is tagged by upper bound of a given bucket.
+--  The bar value corresponds to the number of stakes
+--  In the future the bar value could be different things:
+--  (a) sum of stakes in a bucket
+--  (b) avg or std of stake in a bucket
+--  (c) topN buckets
+--  to name a few
+data HistogramBar = HistogramBarCount
+    { bucketName       :: !Text
+    , bucketUpperBound :: !Word64
+    } deriving (Show, Eq, Ord, Generic)
+
+--  Buckets boundaries can be constructed in different way
+data BoundType = Log10 | Haphazard
+
+generateBounds :: BoundType -> NonEmpty Word64
+generateBounds bType =
+    let (^!) :: Word64 -> Word64 -> Word64
+        (^!) = (^)
+    in case bType of
+        Log10 -> NL.fromList $ ( map (\toPower -> 10 ^! toPower) [1..16] ) ++ [45 * (10 ^! 15)]
+        Haphazard -> NL.fromList [10, 100, 1000, 10000]
+
+instance Arbitrary HistogramBar where
+    arbitrary =
+        let possibleBuckets = fmap show (generateBounds Log10)
+            possibleBars = NL.zipWith HistogramBarCount possibleBuckets (NL.fromList [0..])
+        in elements (NL.toList possibleBars)
+
+
+deriveSafeBuildable ''HistogramBar
+instance BuildableSafeGen HistogramBar where
+    buildSafeGen _ HistogramBarCount{..} =
+        bprint ("{"
+                %" name="%build
+                %" upperBound="%build
+                %" }")
+        bucketName
+        bucketUpperBound
+
+
+data UtxoStatistics = UtxoStatistics
+    { theHistogram :: ![HistogramBar]
+    , theAllStakes :: !Word64
+    } deriving (Show, Generic, Ord)
+
+toMap :: [HistogramBar] -> Map Text Word64
+toMap = Map.fromList . map (\(HistogramBarCount key val) -> (key,val))
+
+instance Eq UtxoStatistics where
+    (UtxoStatistics h s) == (UtxoStatistics h' s') = s == s' && toMap h == toMap h'
+
+instance ToJSON UtxoStatistics where
+    toJSON (UtxoStatistics bars allStakes) =
+        let histogramObject = Object . HMS.fromList . map extractBarKey
+            extractBarKey (HistogramBarCount bound stake) = bound .= stake
+        in object [ "histogram" .= histogramObject bars
+                  , "allStakes" .= allStakes ]
+
+instance FromJSON UtxoStatistics where
+    parseJSON (Object v) =
```

Comment:
It is better to use `Data.Aeson.withObject :: String -> (Object -> Parser a) -> Parser a` rather than pattern matching directly on the constructors. The `String` label is used for error messages, which is really helpful when debugging a parse failure.

So, replace:

```haskell
parseJSON (Object v) = ...
parseJSON _ = empty

-- with:
parseJSON = withObject "UtxoStatistics" $ ...
```

Reply:
refactored as suggested